<a href="https://colab.research.google.com/github/MANISH-KUMAR-CODES/UK_Prime_Minister_Rishi_Sunak-_Twitter_Sentiment_Analysis/blob/main/UK_Prime_Minister_Rishi_Sunak__Twitter_Sentiment_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Dataset Information

Rishi Sunak lost to Liz Truss in September, but she resigned six weeks later. In the latest leadership contest, Mr Sunak racked up the support of his fellow MPs early, and fast. He crossed the 100 nominations he needed long before the deadline - including from MPs that had previously backed Truss or Boris Johnson.
BBC

* Twitter Reaction to Rishi Sunak being elected as PM.
* Sentimental Analysis can be performed on this dataset.

Hashtags: #rishisunak, #rishisunakPM , Rishi Sunak
* Dataset consists of 113578 tweets.

In [45]:
import pandas as pd
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.style as style
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

import re
import string

import nltk
from nltk.probability import FreqDist
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
from nltk.stem import SnowballStemmer
from nltk import pos_tag
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize

from wordcloud import WordCloud
from tqdm.auto import tqdm

In [26]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [27]:
# read the csv file
df = pd.read_csv('/content/drive/MyDrive/uk_pm.csv/uk_pm.csv')
df = pd.DataFrame(df)

## Knowing our dataframe

In [28]:
# checking head of our dataframe
df.head(5)

id                                               text  \
0  1584696134693711873  @BritIndianVoice @RishiSunak #RishiSunak is #B...   
1  1584696057841098752  #RishiSunak is the richest Tory MP in history....   
2  1584695906091536385                                        #RishiSunak   
3  1584695822658441217  A mountain full of money and a knowledge of pr...   
4  1584695640696946688  @trussliz @RishiSunak Chance was been given as...   

          username                                           hashtags  \
0      IsmailYSyed  ['RishiSunak', 'BritishHindu', 'Pakistani', 'P...   
1       AdameMedia                                     ['RishiSunak']   
2  ROHITKU96224077                                     ['RishiSunak']   
3       Parisistas                                     ['RishiSunak']   
4    baboo_prasant                                     ['RishiSunak']   

                  created_at  user followers count  replycount  retweetcount  \
0  2022-10-24 23:59:35+00:00                   393           0             0   
1  2022-10-24 23:59:17+00:00                  9830           3            23   
2  2022-10-24 23:58:40+00:00                     3           0             0   
3  2022-10-24 23:58:21+00:00                     0           0             0   
4  2022-10-24 23:57:37+00:00                    14           0             0   

   likecount  quotecount language  quotedtweet  inReplyToTweetId  \
0          0           0       en          NaN      1.584541e+18   
1         68           4       en          NaN               NaN   
2          0           0      qht          NaN               NaN   
3          0           0       en          NaN               NaN   
4          0           0       en          NaN      1.584546e+18   

                         inReplyToUser  \
0  https://twitter.com/BritIndianVoice   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4         https://twitter.com/trussliz   

                                      mentionedUsers  
0  @BritIndianVoice @RishiSunak #RishiSunak is #B...  
1  #RishiSunak is the richest Tory MP in history....  
2                                        #RishiSunak  
3  A mountain full of money and a knowledge of pr...  
4  @trussliz @RishiSunak Chance was been given as...

In [29]:
#checking tail of dataframe
df.tail()

id  \
113573  1584553348736421888   
113574  1584553348677525505   
113575  1584553348597895168   
113576  1584553348132397056   
113577  1584553347998105600   

                                                     text       username  \
113573  @JackBarbour_ @piersmorgan @RishiSunak The man...     aqeelone22   
113574  @Jacob_Rees_Mogg @RishiSunak Let's hope you're...  EddieThornley   
113575  @sajidjavid @RishiSunak Desperate call for a d...    pedleysjohn   
113576  @SkyNews Rulling Party needs a strong oppositi...       KarobiVC   
113577  @trussliz @RishiSunak Liz Truss has won the ma...     sands_hill   

                       hashtags                 created_at  \
113573                      NaN  2022-10-24 14:32:12+00:00   
113574                      NaN  2022-10-24 14:32:12+00:00   
113575  ['ToriesUnfitToGovern']  2022-10-24 14:32:12+00:00   
113576                      NaN  2022-10-24 14:32:12+00:00   
113577                      NaN  2022-10-24 14:32:12+00:00   

        user followers count  replycount  retweetcount  likecount  quotecount  \
113573                  1162           0             0          0           0   
113574                    97           0             0          0           0   
113575                  2484           0             0          0           0   
113576                   392           0             0          0           0   
113577                     4           0             0          0           0   

       language  quotedtweet  inReplyToTweetId  \
113573       en          NaN      1.584532e+18   
113574       en          NaN      1.584539e+18   
113575       en          NaN      1.584401e+18   
113576       en          NaN      1.584512e+18   
113577       en          NaN      1.584546e+18   

                              inReplyToUser  \
113573     https://twitter.com/JackBarbour_   
113574  https://twitter.com/Jacob_Rees_Mogg   
113575       https://twitter.com/sajidjavid   
113576          https://twitter.com/SkyNews   
113577         https://twitter.com/trussliz   

                                           mentionedUsers  
113573  @JackBarbour_ @piersmorgan @RishiSunak The man...  
113574  @Jacob_Rees_Mogg @RishiSunak Let's hope you're...  
113575  @sajidjavid @RishiSunak Desperate call for a d...  
113576  @SkyNews Rulling Party needs a strong oppositi...  
113577  @trussliz @RishiSunak Liz Truss has won the ma...

In [30]:
# checking info of our dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113578 entries, 0 to 113577
Data columns (total 15 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   id                    113578 non-null  int64  
 1   text                  113578 non-null  object 
 2   username              113578 non-null  object 
 3   hashtags              26957 non-null   object 
 4   created_at            113578 non-null  object 
 5   user followers count  113578 non-null  int64  
 6   replycount            113578 non-null  int64  
 7   retweetcount          113578 non-null  int64  
 8   likecount             113578 non-null  int64  
 9   quotecount            113578 non-null  int64  
 10  language              113578 non-null  object 
 11  quotedtweet           0 non-null       float64
 12  inReplyToTweetId      83368 non-null   float64
 13  inReplyToUser         83368 non-null   object 
 14  mentionedUsers        113578 non-null  object 
dtype

### From above we can see that our dataframe countains 113578 rows and 15 columns out of which we have a column ('quotedtweet') with zero entires, its better we drop this column

In [31]:
# check if text column is same as mentionedusers
sum = 0
for i in range(0,len(df)):
    if df['text'][i]!=df['mentionedUsers'][i]:
       sum = sum+1

print(sum)

0


### from the code above we find out that our text column is same as our mentionedUsers columns ,so lets drop it also.


In [32]:
#dropppping mentionedUsers column from our dataframe
df = df.drop('mentionedUsers',axis = 1)
df.head(5)

id                                               text  \
0  1584696134693711873  @BritIndianVoice @RishiSunak #RishiSunak is #B...   
1  1584696057841098752  #RishiSunak is the richest Tory MP in history....   
2  1584695906091536385                                        #RishiSunak   
3  1584695822658441217  A mountain full of money and a knowledge of pr...   
4  1584695640696946688  @trussliz @RishiSunak Chance was been given as...   

          username                                           hashtags  \
0      IsmailYSyed  ['RishiSunak', 'BritishHindu', 'Pakistani', 'P...   
1       AdameMedia                                     ['RishiSunak']   
2  ROHITKU96224077                                     ['RishiSunak']   
3       Parisistas                                     ['RishiSunak']   
4    baboo_prasant                                     ['RishiSunak']   

                  created_at  user followers count  replycount  retweetcount  \
0  2022-10-24 23:59:35+00:00                   393           0             0   
1  2022-10-24 23:59:17+00:00                  9830           3            23   
2  2022-10-24 23:58:40+00:00                     3           0             0   
3  2022-10-24 23:58:21+00:00                     0           0             0   
4  2022-10-24 23:57:37+00:00                    14           0             0   

   likecount  quotecount language  quotedtweet  inReplyToTweetId  \
0          0           0       en          NaN      1.584541e+18   
1         68           4       en          NaN               NaN   
2          0           0      qht          NaN               NaN   
3          0           0       en          NaN               NaN   
4          0           0       en          NaN      1.584546e+18   

                         inReplyToUser  
0  https://twitter.com/BritIndianVoice  
1                                  NaN  
2                                  NaN  
3                                  NaN  
4         https://twitter.com/trussliz

In [33]:
df = df.drop('quotedtweet',axis = 1)
# also lets check the null values present in our dataframe
df.isnull().sum()

id                          0
text                        0
username                    0
hashtags                86621
created_at                  0
user followers count        0
replycount                  0
retweetcount                0
likecount                   0
quotecount                  0
language                    0
inReplyToTweetId        30210
inReplyToUser           30210
dtype: int64

### From above we found that our hastag columns has 86621 null values,inReplyToTweetId and inReplyToUser columns have 30210 null values

###To know the statistical summary of the DataFrame columns, we use the describe() function in python. And to know the total number of rows and columns of the dataset, we use the .shape function in python.

In [53]:
# checking description of our dataset
df.describe()

id  user followers count     replycount   retweetcount  \
count  1.135780e+05          1.135780e+05  113578.000000  113578.000000   
mean   1.584272e+18          5.732325e+03       0.990456       1.824843   
std    2.161895e+15          2.808652e+05      32.213791      91.932817   
min    1.565130e+18          0.000000e+00       0.000000       0.000000   
25%    1.584564e+18          3.500000e+01       0.000000       0.000000   
50%    1.584590e+18          2.030000e+02       0.000000       0.000000   
75%    1.584623e+18          9.780000e+02       0.000000       0.000000   
max    1.584696e+18          8.371055e+07    5177.000000   25053.000000   

           likecount     quotecount  inReplyToTweetId  
count  113578.000000  113578.000000      8.336800e+04  
mean       11.951003       0.153833      1.584147e+18  
std       662.464267       6.997948      7.023966e+15  
min         0.000000       0.000000      1.120754e+18  
25%         0.000000       0.000000      1.584531e+18  
50%         0.000000       0.000000      1.584541e+18  
75%         1.000000       0.000000      1.584566e+18  
max    185559.000000    1115.000000      1.584695e+18

### For our sentiment analysis we need two important 'ID' and 'Text' columns

In [34]:
df1 = df[['id','text']]
df1.head(5)

id                                               text
0  1584696134693711873  @BritIndianVoice @RishiSunak #RishiSunak is #B...
1  1584696057841098752  #RishiSunak is the richest Tory MP in history....
2  1584695906091536385                                        #RishiSunak
3  1584695822658441217  A mountain full of money and a knowledge of pr...
4  1584695640696946688  @trussliz @RishiSunak Chance was been given as...

## Preprocessing the dataset

In [35]:
# removes pattern in the text column]
def remove_pattern(input_text,pattern):
    r = re.findall(pattern,input_text)
    for word in r:
       input_text = re.sub(word,'',input_text)
    return input_text   

In [36]:
# remove twitter handles @user
df1['clean_tweet'] = np.vectorize(remove_pattern)(df1['text'], "@[\w]*")
df1['clean_tweet'][0]

'  #RishiSunak is #BritishHindu of paternal #Pakistani Punjabi origin. He’s legally POC (#Pakistan\xa0Origin Card) holder issued by  (probably also concurrently holding India’s OCI through his Indian NRI wife/mother). Also there may be issue with his wife’s Indian citizenship.'

In [37]:
# remove special characters, numbers and punctuations
df1['clean_tweet'] = df1['clean_tweet'].str.replace("[^a-zA-Z#]", " ")
df1['clean_tweet'][0]

'  #RishiSunak is #BritishHindu of paternal #Pakistani Punjabi origin  He s legally POC  #Pakistan Origin Card  holder issued by   probably also concurrently holding India s OCI through his Indian NRI wife mother   Also there may be issue with his wife s Indian citizenship '

In [48]:
# remove short words
df1['clean_tweet'] = df1['clean_tweet'].apply(lambda x: " ".join([w for w in x.split() if len(w)>3]))
df1['clean_tweet'][0]

'#RishiSunak #BritishHindu paternal #Pakistani Punjabi origin legally #Pakistan Origin Card holder issued probably also concurrently holding India through Indian wife mother Also there issue with wife Indian citizenship'

In [49]:
# individual words considered as tokens
tokenized_tweet = df1['clean_tweet'].apply(lambda x: x.split())
tokenized_tweet.head()

0    [#RishiSunak, #BritishHindu, paternal, #Pakist...
1    [#RishiSunak, richest, Tory, history, wife, bi...
2                                        [#RishiSunak]
3    [mountain, full, money, knowledge, proven, the...
4    [Chance, been, given, didn, fulfilled, duty, p...
Name: clean_tweet, dtype: object

##Stemming words with NLTK

Stemming is the process of producing morphological variants of a root/base word. Stemming programs are commonly referred to as stemming algorithms or stemmers. A stemming algorithm reduces the words “chocolates”, “chocolatey”, and “choco” to the root word, “chocolate” and “retrieval”, “retrieved”, “retrieves” reduce to the stem “retrieve”.

In [50]:
# stem the words
from nltk.stem.porter import PorterStemmer
stemmer = PorterStemmer()

tokenized_tweet = tokenized_tweet.apply(lambda sentence: [stemmer.stem(word) for word in sentence])
tokenized_tweet.head()

0    [#rishisunak, #britishhindu, patern, #pakistan...
1    [#rishisunak, richest, tori, histori, wife, bi...
2                                        [#rishisunak]
3    [mountain, full, money, knowledg, proven, theo...
4    [chanc, been, given, didn, fulfil, duti, peopl...
Name: clean_tweet, dtype: object

In [51]:
# combine words into single sentence
for i in range(len(tokenized_tweet)):
    tokenized_tweet[i] = " ".join(tokenized_tweet[i])
    
df1['clean_tweet'] = tokenized_tweet
df1.head()

id                                               text  \
0  1584696134693711873  @BritIndianVoice @RishiSunak #RishiSunak is #B...   
1  1584696057841098752  #RishiSunak is the richest Tory MP in history....   
2  1584695906091536385                                        #RishiSunak   
3  1584695822658441217  A mountain full of money and a knowledge of pr...   
4  1584695640696946688  @trussliz @RishiSunak Chance was been given as...   

                                         clean_tweet  
0  #rishisunak #britishhindu patern #pakistani pu...  
1  #rishisunak richest tori histori wife billiona...  
2                                        #rishisunak  
3  mountain full money knowledg proven theori wit...  
4  chanc been given didn fulfil duti peopl late #...

In [52]:
# checking the first item of our clean tweet
df1['clean_tweet'][0]

'#rishisunak #britishhindu patern #pakistani punjabi origin legal #pakistan origin card holder issu probabl also concurr hold india through indian wife mother also there issu with wife indian citizenship'